src/Microsoft.Diagnostics.EventFlow.Core/Implementations/PayloadDictionaryUtilities.cs

```diff
@@ -16,24 +16,40 @@ public static void AddPayloadProperty(IDictionary<string, object> payload, strin
             Requires.NotNull(key, nameof(key));
             Requires.NotNull(healthReporter, nameof(healthReporter));
 
-            if (!payload.ContainsKey(key))
+            if (!payload.TryGetValue(key, out var existingValue))
             {
                 payload.Add(key, value);
                 return;
             }
+            else if ((existingValue?.Equals(value)).GetValueOrDefault(false))
+            {
+                // Existing value with same key is equivalent to the input value
+                // We can return immediately to avoid adding duplicate key/value into the payload
+                healthReporter.ReportWarning($"The property with the key '{key}' already exist in the event payload with equivalent value. Value was not re-added", context);
```

Comment:
Do not warn, just return. Emitting a warning has operational implications (the relevant component might be considered to be in "warning" state, requiring human intervention). In this case we might be trying to add some data more than once, but there is no data loss, so we still should be considered generally "healthy"